pkg/controller/cni/cilium_reconciler.go

```diff
@@ -0,0 +1,231 @@
+	"context"
+	"errors"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	eksacluster "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/controller"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+	"github.com/aws/eks-anywhere/pkg/executables"
+	"github.com/aws/eks-anywhere/pkg/networking/cilium"
+)
+
+const defaultRequeueTime = time.Second * 10
+
+type CiliumReconciler struct {
+	tracker *remote.ClusterCacheTracker
+}
+
+func (cr *CiliumReconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *anywherev1.Cluster, capiCluster *clusterv1.Cluster, specWithBundles *eksacluster.Spec) (controller.Result, error) {
+	log.Info("Getting remote client", "client for cluster", capiCluster.Name)
+	key := client.ObjectKey{
+		Namespace: capiCluster.Namespace,
+		Name:      capiCluster.Name,
+	}
+	remoteClient, err := cr.tracker.GetClient(ctx, key)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	log.Info("Applying CNI")
+	ciliumDS, err := getCiliumDS(ctx, remoteClient)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Deploying Cilium DS")
+			helm := executables.NewHelm(executables.NewExecutable("helm"), executables.WithInsecure())
+
+			ci := cilium.NewCilium(nil, helm)
+
+			ciliumSpec, err := ci.GenerateManifest(ctx, specWithBundles, []string{constants.CapvSystemNamespace})
+			if err != nil {
+				return controller.Result{}, err
+			}
+			if err := serverside.ReconcileYaml(ctx, remoteClient, ciliumSpec); err != nil {
+				return controller.Result{}, err
+			}
+			return controller.Result{}, err
+		}
+
+		return controller.Result{}, err
+	}
+
+	// upgrade cilium
+	log.Info("Upgrading Cilium")
+	needsUpgrade, err := ciliumNeedsUpgrade(ctx, log, remoteClient, specWithBundles)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	if !needsUpgrade {
+		log.Info("Skipping Cilium")
+		return controller.Result{}, nil
+	}
+
+	helm := executables.NewHelm(executables.NewExecutable("helm"))
+	templater := cilium.NewTemplater(helm)
+	preflight, err := templater.GenerateUpgradePreflightManifest(ctx, specWithBundles)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	log.Info("Installing Cilium upgrade preflight manifest")
+	if err := serverside.ReconcileYaml(ctx, remoteClient, preflight); err != nil {
+		return controller.Result{}, err
+	}
+
+	if err := cilium.CheckDaemonSetReady(ciliumDS); err != nil {
+		log.Error(err, "error checking if Cilium DaemonSet is ready")
+		return controller.Result{Result: &ctrl.Result{
+			RequeueAfter: defaultRequeueTime,
+		}}, nil
```

Comment:
Should we be returning err here?

Reply:
The goal here is to reschedule reconciliation after defaultRequeueTime, but the current code returns an empty ctrl.Result if an error is returned by clusterProviderReconciler.Reconcile()

https://github.com/aws/eks-anywhere/blob/main/controllers/cluster_controller.go#L164